src/main/java/edu/kit/datamanager/security/filter/KeycloakTokenFilter.java

```diff
@@ -0,0 +1,181 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * This original version of this code is available at 
+ *
+ *  https://github.com/akoserwal/keycloak-jwt 
+ * 
+ * and was modified according to our requirements. 
+ */
+package edu.kit.datamanager.security.filter;
+
+import com.nimbusds.jose.proc.BadJOSEException;
+import edu.kit.datamanager.exceptions.InvalidAuthenticationException;
+import edu.kit.datamanager.util.NetworkUtils;
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.ExpiredJwtException;
+import io.jsonwebtoken.Jws;
+import io.jsonwebtoken.MalformedJwtException;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.web.filter.OncePerRequestFilter;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author akoserwa@redhat.com
+ */
+public class KeycloakTokenFilter extends OncePerRequestFilter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KeycloakTokenFilter.class);
+
+    private static final String BEARER = "Bearer ";
+    public final static String AUTHORIZATION_HEADER = "Authorization";
+    private final KeycloakTokenValidator tokenValidator;
+    private JwtAuthenticationToken jwToken;
+
+    /**
+     * Default constructor.
+     *
+     * @param tokenValidator The token validator
+     */
+    public KeycloakTokenFilter(KeycloakTokenValidator tokenValidator) {
+        this.tokenValidator = tokenValidator;
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+        String token = request.getHeader(AUTHORIZATION_HEADER);
+        if (token != null && !token.toUpperCase().startsWith("BASIC") && token.startsWith(BEARER)) {
+            LOG.trace("Starting JWT filtering.");
+            try {
+                //remove Bearer prefix
+                if (token.length() > BEARER.length()) {
+                    token = token.substring(BEARER.length());
+                }
+
+                if (tokenValidator.supportsLocalJwt()) {
```

Comment:
filterChain.doFilter is code duplication
do
if (tokenValidator.supportsLocalJwt() && attemptLocalAuthentication(request, response, token)) {
                       LOG.debug("Authenticated using local JWT secret.");
 } else {
...
}
filterChain.doFilter(request, response);

Reply:
I don't like return in the middle of the method.